File: src/WorkBC.ElasticSearch.Models/Filters/JobSearchFilters.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using WorkBC.ElasticSearch.Models.JobAttributes;

namespace WorkBC.ElasticSearch.Models.Filters
{
    public class JobSearchFilters : IPageableFilters
    {
        public DateField StartDate { get; set; }

        public DateField EndDate { get; set; }

        /// <example>1</example>
        public int Page { get; set; }

        /// <example>20</example>
        public int PageSize { get; set; }

        /// <example></example>
        public string Keyword { get; set; }

        /// <example>all</example>
        public string SearchInField { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeFullTime { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypePartTime { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeLeadingToFullTime { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypePermanent { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeTemporary { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeCasual { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeSeasonal { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeDay { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeEarly { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeEvening { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeFlexible { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeMorning { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeNight { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeOnCall { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeOvertime { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeShift { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeTbd { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeWeekend { get; set; }

        /// <summary>
        ///     ANY_DATE = "0", TODAY = "1", PAST_THREE_DAYS = "2", DATE_RANGE = "3"
        /// </summary>
        /// <example>0</example>
        public string SearchDateSelection { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeOnSite { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeHybrid { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeTravelling { get; set; }

        [DefaultValue(false)]
        public bool SearchJobTypeVirtual { get; set; }

        /// <example>["University", "College or apprenticeship", "Secondary school or job-specific training", "No education"]</example>
        public List<string> SearchJobEducationLevel { get; set; }

        /// <example>4</example>
        public int SalaryType { get; set; }

        /// <example>-1</example>
        public int SearchLocationDistance { get; set; }

        public List<LocationField> SearchLocations { get; set; }

        /// <example>
        ///     ["As per collective agreement", "Bonus", "Commission", "Dental benefits", "Disability benefits", "Gratuities",
        ///     "Group insurance benefits", "Life insurance benefits", "Medical benefits", "Mileage paid", "Pension plan benefits",
        ///     "Piece work", "RESP benefits", "RRSP benefits", "Vision care benefits", "Other benefits"]
        /// </example>
        public List<string> SearchSalaryConditions { get; set; }

        /// <example>11</example>
        public int SortOrder { get; set; }

        [DefaultValue(false)]
        public bool SalaryBracket1 { get; set; }

        [DefaultValue(false)]
        public bool SalaryBracket2 { get; set; }

        [DefaultValue(false)]
        public bool SalaryBracket3 { get; set; }

        [DefaultValue(false)]
        public bool SalaryBracket4 { get; set; }

        [DefaultValue(false)]
        public bool SalaryBracket5 { get; set; }

        [DefaultValue(false)]
        public bool SalaryBracket6 { get; set; }
        
        /// <example></example>
        public string SalaryMin { get; set; }

        /// <example></example>
        public string SalaryMax { get; set; }

        /// <example>
        ///     ["37", "40", "1", "36", "23", "34", "42", "29", "35", "28", "32", "24", "21", "44", "46", "31", "39", "30",
        ///     "45", "43", "26", "27", "22", "41", "25" ]
        /// </example>
        public List<int> SearchIndustry { get; set; }

        [DefaultValue(false)]
        public bool SearchIsApprentice { get; set; }

        [DefaultValue(false)]
        public bool SearchIsVeterans { get; set; }

        [DefaultValue(false)]
        public bool SearchIsIndigenous { get; set; }

        [DefaultValue(false)]
        public bool SearchIsMatureWorkers { get; set; }

        [DefaultValue(false)]
        public bool SearchIsNewcomers { get; set; }

        [DefaultValue(false)]
        public bool SearchIsPeopleWithDisabilities { get; set; }

        [DefaultValue(false)]
        public bool SearchIsStudents { get; set; }

        [DefaultValue(false)]
        public bool SearchIsVisibleMinority { get; set; }

        [DefaultValue(false)]
        public bool SearchIsYouth { get; set; }

        [DefaultValue(true)]
        public bool SearchIsPostingsInEnglish { get; set; }

        [DefaultValue(false)]
        public bool SearchIsPostingsInEnglishAndFrench { get; set; }

        /// <example></example>
        public string SearchNocField { get; set; }

        /// <example>0</example>
        public string SearchJobSource { get; set; }

        [DefaultValue(false)]
        public bool SearchExcludePlacementAgencyJobs { get; set; }
    }

    public class DateField
    {
        public DateField()
        {
            Year = 0;
            Month = 0;
            Day = 0;
            Hour = 0;
            Minute = 0;
            Second = 0;
            Millisecond = 0;
        }

        public DateField(DateTime date)
        {
            Year = date.Year;
            Month = date.Month;
            Day = date.Day;
            Hour = date.Hour;
            Minute = date.Minute;
            Second = date.Second;
            Millisecond = date.Millisecond;
        }

        public int Year { get; set; }
        public int Month { get; set; }
        public int Day { get; set; }
        public int Hour { get; set; }
        public int Minute { get; set; }
        public int Second { get; set; }
        public int Millisecond { get; set; }

        // returns a YYYY-MM-DD string
        public override string ToString()
        {
            return $"{Year}-{Month:00}-{Day:00}T{Hour:00}:{Minute:00}:{Second:00}.{Millisecond:000}";
        }
    }

    public class LocationField
    {
        private string _postal;

        /// <example>Surrey</example>
        public string City { get; set; }

        /// <example></example>
        public string Region { get; set; }

        /// <example></example>
        public string Postal
        {
            // uppercase postal codes and remove spaces
            get => _postal?.ToUpper().Replace(" ", "");
            set => _postal = value;
        }
    }
}

File: src/WorkBC.Notifications.JobAlerts.V2/src/Search/JobSearchFilters.php
<?php

declare(strict_types=1);

namespace WorkBC\JobAlertNotifier\Search;

/**
 * Port of WorkBC.ElasticSearch.Models.Filters.JobSearchFilters (plus the
 * nested DateField / LocationField types).
 *
 * The JobAlerts.JobSearchFilters column stores the JSON the Angular SPA
 * produced when the alert was saved (PascalCase keys). The C# task
 * deserializes it with Newtonsoft, which matches property names
 * case-insensitively and leaves missing properties at their CLR defaults —
 * fromJson() mirrors both behaviours.
 */
final class JobSearchFilters
{
    public ?DateField $startDate = null;
    public ?DateField $endDate = null;
    public int $page = 0;
    public int $pageSize = 0;
    public ?string $keyword = null;
    public ?string $searchInField = null;

    public bool $searchJobTypeFullTime = false;
    public bool $searchJobTypePartTime = false;
    public bool $searchJobTypeLeadingToFullTime = false;
    public bool $searchJobTypePermanent = false;
    public bool $searchJobTypeTemporary = false;
    public bool $searchJobTypeCasual = false;
    public bool $searchJobTypeSeasonal = false;
    public bool $searchJobTypeDay = false;
    public bool $searchJobTypeEarly = false;
    public bool $searchJobTypeEvening = false;
    public bool $searchJobTypeFlexible = false;
    public bool $searchJobTypeMorning = false;
    public bool $searchJobTypeNight = false;
    public bool $searchJobTypeOnCall = false;
    public bool $searchJobTypeOvertime = false;
    public bool $searchJobTypeShift = false;
    public bool $searchJobTypeTbd = false;
    public bool $searchJobTypeWeekend = false;
    public bool $searchJobTypeOnSite = false;
    public bool $searchJobTypeHybrid = false;
    public bool $searchJobTypeTravelling = false;
    public bool $searchJobTypeVirtual = false;

    /** ANY_DATE = "0", TODAY = "1", PAST_THREE_DAYS = "2", DATE_RANGE = "3" */
    public ?string $searchDateSelection = null;

    /** @var list<string>|null */
    public ?array $searchJobEducationLevel = null;

    public int $salaryType = 0;
    public int $searchLocationDistance = 0;

    /** @var list<LocationField>|null */
    public ?array $searchLocations = null;

    /** @var list<string>|null */
    public ?array $searchSalaryConditions = null;

    public int $sortOrder = 0;

    public bool $salaryBracket1 = false;
    public bool $salaryBracket2 = false;
    public bool $salaryBracket3 = false;
    public bool $salaryBracket4 = false;
    public bool $salaryBracket5 = false;
    public bool $salaryBracket6 = false;
    public bool $searchSalaryUnknown = false;
    public ?string $salaryMin = null;
    public ?string $salaryMax = null;

    /** @var list<int>|null */
    public ?array $searchIndustry = null;

    public bool $searchIsApprentice = false;
    public bool $searchIsVeterans = false;
    public bool $searchIsIndigenous = false;
    public bool $searchIsMatureWorkers = false;
    public bool $searchIsNewcomers = false;
    public bool $searchIsPeopleWithDisabilities = false;
    public bool $searchIsStudents = false;
    public bool $searchIsVisibleMinority = false;
    public bool $searchIsYouth = false;
    public bool $searchIsPostingsInEnglish = false;
    public bool $searchIsPostingsInEnglishAndFrench = false;

    public ?string $searchNocField = null;
    public ?string $searchJobSource = null;
    public bool $searchExcludePlacementAgencyJobs = false;

    public static function fromJson(string $json): self
    {
        $data = json_decode($json, true);
        if (!is_array($data)) {
            throw new \InvalidArgumentException('JobSearchFilters column does not contain valid JSON');
        }
        return self::fromArray($data);
    }

    /** @param array<string, mixed> $data */
    public static function fromArray(array $data): self
    {
        // Newtonsoft matches property names case-insensitively.
        $map = [];
        foreach ($data as $key => $value) {
            $map[strtolower((string) $key)] = $value;
        }

        $f = new self();

        $f->startDate = DateField::fromValue($map['startdate'] ?? null);
        $f->endDate = DateField::fromValue($map['enddate'] ?? null);
        $f->page = self::toInt($map['page'] ?? 0);
        $f->pageSize = self::toInt($map['pagesize'] ?? 0);
        $f->keyword = self::toStringOrNull($map['keyword'] ?? null);
        $f->searchInField = self::toStringOrNull($map['searchinfield'] ?? null);

        foreach ([
            'searchJobTypeFullTime', 'searchJobTypePartTime', 'searchJobTypeLeadingToFullTime',
            'searchJobTypePermanent', 'searchJobTypeTemporary', 'searchJobTypeCasual',
            'searchJobTypeSeasonal', 'searchJobTypeDay', 'searchJobTypeEarly',
            'searchJobTypeEvening', 'searchJobTypeFlexible', 'searchJobTypeMorning',
            'searchJobTypeNight', 'searchJobTypeOnCall', 'searchJobTypeOvertime',
            'searchJobTypeShift', 'searchJobTypeTbd', 'searchJobTypeWeekend',
            'searchJobTypeOnSite', 'searchJobTypeHybrid', 'searchJobTypeTravelling',
            'searchJobTypeVirtual',
            'salaryBracket1', 'salaryBracket2', 'salaryBracket3', 'salaryBracket4',
            'salaryBracket5', 'salaryBracket6', 'searchSalaryUnknown',
            'searchIsApprentice', 'searchIsVeterans', 'searchIsIndigenous',
            'searchIsMatureWorkers', 'searchIsNewcomers', 'searchIsPeopleWithDisabilities',
            'searchIsStudents', 'searchIsVisibleMinority', 'searchIsYouth',
            'searchIsPostingsInEnglish', 'searchIsPostingsInEnglishAndFrench',
            'searchExcludePlacementAgencyJobs',
        ] as $prop) {
            $f->{$prop} = self::toBool($map[strtolower($prop)] ?? false);
        }

        $f->searchDateSelection = self::toStringOrNull($map['searchdateselection'] ?? null);
        $f->searchJobEducationLevel = self::toStringList($map['searchjobeducationlevel'] ?? null);
        $f->salaryType = self::toInt($map['salarytype'] ?? 0);
        $f->searchLocationDistance = self::toInt($map['searchlocationdistance'] ?? 0);
        $f->searchSalaryConditions = self::toStringList($map['searchsalaryconditions'] ?? null);
        $f->sortOrder = self::toInt($map['sortorder'] ?? 0);
        $f->salaryMin = self::toStringOrNull($map['salarymin'] ?? null);
        $f->salaryMax = self::toStringOrNull($map['salarymax'] ?? null);
        $f->searchNocField = self::toStringOrNull($map['searchnocfield'] ?? null);
        $f->searchJobSource = self::toStringOrNull($map['searchjobsource'] ?? null);

        $industries = $map['searchindustry'] ?? null;
        if (is_array($industries)) {
            $f->searchIndustry = array_values(array_map(fn($v) => self::toInt($v), $industries));
        }

        $locations = $map['searchlocations'] ?? null;
        if (is_array($locations)) {
            $f->searchLocations = array_values(array_map(
                fn($loc) => LocationField::fromValue($loc),
                $locations
            ));
        }

        return $f;
    }

    private static function toBool(mixed $value): bool
    {
        if (is_bool($value)) {
            return $value;
        }
        if (is_string($value)) {
            return strtolower($value) === 'true' || $value === '1';
        }
        return (bool) $value;
    }

    private static function toInt(mixed $value): int
    {
        if (is_int($value)) {
            return $value;
        }
        if (is_numeric($value)) {
            return (int) $value;
        }
        return 0;
    }

    private static function toStringOrNull(mixed $value): ?string
    {
        if ($value === null) {
            return null;
        }
        if (is_scalar($value)) {
            return (string) $value;
        }
        return null;
    }

    /** @return list<string>|null */
    private static function toStringList(mixed $value): ?array
    {
        if (!is_array($value)) {
            return null;
        }
        return array_values(array_map(fn($v) => (string) $v, $value));
    }
}

/**
 * Port of the C# DateField. ToString() renders the same
 * "{Year}-{Month:00}-{Day:00}T{Hour:00}:{Minute:00}:{Second:00}.{Millisecond:000}"
 * format used inside the DatePosted range filter.
 */
final class DateField
{
    public int $year = 0;
    public int $month = 0;
    public int $day = 0;
    public int $hour = 0;
    public int $minute = 0;
    public int $second = 0;
    public int $millisecond = 0;

    public static function fromDateTime(\DateTimeInterface $date): self
    {
        $d = new self();
        $d->year = (int) $date->format('Y');
        $d->month = (int) $date->format('n');
        $d->day = (int) $date->format('j');
        $d->hour = (int) $date->format('G');
        $d->minute = (int) $date->format('i');
        $d->second = (int) $date->format('s');
        $d->millisecond = (int) $date->format('v');
        return $d;
    }

    public static function fromValue(mixed $value): ?self
    {
        if (!is_array($value)) {
            return null;
        }

        $map = [];
        foreach ($value as $key => $v) {
            $map[strtolower((string) $key)] = $v;
        }

        $d = new self();
        $d->year = (int) ($map['year'] ?? 0);
        $d->month = (int) ($map['month'] ?? 0);
        $d->day = (int) ($map['day'] ?? 0);
        $d->hour = (int) ($map['hour'] ?? 0);
        $d->minute = (int) ($map['minute'] ?? 0);
        $d->second = (int) ($map['second'] ?? 0);
        $d->millisecond = (int) ($map['millisecond'] ?? 0);
        return $d;
    }

    public function __toString(): string
    {
        return sprintf(
            '%d-%02d-%02dT%02d:%02d:%02d.%03d',
            $this->year, $this->month, $this->day,
            $this->hour, $this->minute, $this->second, $this->millisecond
        );
    }
}

/**
 * Port of the C# LocationField. The Postal getter uppercases and strips
 * spaces, exactly like the C# property.
 */
final class LocationField
{
    public ?string $city = null;
    public ?string $region = null;
    private ?string $postal = null;

    public static function fromValue(mixed $value): self
    {
        $l = new self();
        if (!is_array($value)) {
            return $l;
        }

        $map = [];
        foreach ($value as $key => $v) {
            $map[strtolower((string) $key)] = $v;
        }

        $l->city = isset($map['city']) && is_scalar($map['city']) ? (string) $map['city'] : null;
        $l->region = isset($map['region']) && is_scalar($map['region']) ? (string) $map['region'] : null;
        $l->postal = isset($map['postal']) && is_scalar($map['postal']) ? (string) $map['postal'] : null;
        return $l;
    }

    public function postal(): ?string
    {
        if ($this->postal === null) {
            return null;
        }
        return str_replace(' ', '', strtoupper($this->postal));
    }
}
